Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs

```diff
@@ -0,0 +1,263 @@
+            {
+                throw new InvalidOperationException("Error deserializing JSON credential data.", e);
+            }
+            return CreateDefaultCredentialFromJson(credentialParameters);
+        }
+
+        /// <summary>Creates a default credential from JSON data.</summary>
+        private static GoogleCredential CreateDefaultCredentialFromJson(JsonCredentialParameters credentialParameters)
+        {
+            switch (credentialParameters.Type)
+            {
+                case JsonCredentialParameters.AuthorizedUserCredentialType:
+                    return GoogleCredential.FromCredential(CreateUserCredentialFromJson(credentialParameters));
+
+                case JsonCredentialParameters.ServiceAccountCredentialType:
+                    return GoogleCredential.FromCredential(CreateServiceAccountCredentialFromJson(credentialParameters));
+                
+                default:
+                    throw new InvalidOperationException(String.Format("Error creating credential from JSON. Unrecognized credential type {0}.", credentialParameters.Type));
+            }
+        }
+
+        /// <summary>Creates user credential from JSON data.</summary>
+        private static UserCredential CreateUserCredentialFromJson(JsonCredentialParameters credentialParameters)
+        {
+            if (credentialParameters.Type != JsonCredentialParameters.AuthorizedUserCredentialType ||
+                string.IsNullOrEmpty(credentialParameters.ClientId) ||
+                string.IsNullOrEmpty(credentialParameters.ClientSecret))
+            {
+                throw new InvalidOperationException("JSON data do not represent a valid user credential.");
+            }
+
+            var token = new TokenResponse
+            {
+                RefreshToken = credentialParameters.RefreshToken
+            };
+
+            var initializer = new GoogleAuthorizationCodeFlow.Initializer
+            {
+                ClientSecrets = new ClientSecrets
+                {
+                    ClientId = credentialParameters.ClientId,
+                    ClientSecret = credentialParameters.ClientSecret
+                }
+            };
+            var flow = new GoogleAuthorizationCodeFlow(initializer);
+            return new UserCredential(flow, "ApplicationDefaultCredentials", token);
+        }
+
+        /// <summary>Creates service account credential from JSON data.</summary>
+        private static ServiceAccountCredential CreateServiceAccountCredentialFromJson(JsonCredentialParameters credentialParameters)
+        {
+            if (credentialParameters.Type != JsonCredentialParameters.ServiceAccountCredentialType ||
+                string.IsNullOrEmpty(credentialParameters.ClientEmail) ||
+                string.IsNullOrEmpty(credentialParameters.PrivateKey))
+            {
+                throw new InvalidOperationException("JSON data do not represent a valid service account credential.");
```

Comment:
Actually "data" is plural in english (the reason for that being that the word "data" is plural in Latin originally, which sounds a bit strange, but indeed - In latin "data" basically means "things that have been given").